File: packages/go-sdk/fsm/workflow.go
package fsm

import "fmt"

// WorkflowRunStatus represents the status of a workflow run.
type WorkflowRunStatus string

const (
	WorkflowRunPending   WorkflowRunStatus = "pending"
	WorkflowRunRunning   WorkflowRunStatus = "running"
	WorkflowRunPaused    WorkflowRunStatus = "paused"
	WorkflowRunCompleted WorkflowRunStatus = "completed"
	WorkflowRunFailed    WorkflowRunStatus = "failed"
	WorkflowRunTimedOut  WorkflowRunStatus = "timed_out"
	WorkflowRunCanceled  WorkflowRunStatus = "canceled"
)

// WorkflowRunEvent represents an event that triggers a workflow run state transition.
type WorkflowRunEvent string

const (
	WorkflowRunEventStart    WorkflowRunEvent = "START"
	WorkflowRunEventPause    WorkflowRunEvent = "PAUSE"
	WorkflowRunEventResume   WorkflowRunEvent = "RESUME"
	WorkflowRunEventComplete WorkflowRunEvent = "COMPLETE"
	WorkflowRunEventFail     WorkflowRunEvent = "FAIL"
	WorkflowRunEventTimeout  WorkflowRunEvent = "TIMEOUT"
	WorkflowRunEventCancel   WorkflowRunEvent = "CANCEL"
)

var workflowRunTransitions = map[WorkflowRunStatus]map[WorkflowRunEvent]WorkflowRunStatus{
	WorkflowRunPending: {
		WorkflowRunEventStart:  WorkflowRunRunning,
		WorkflowRunEventCancel: WorkflowRunCanceled,
	},
	WorkflowRunRunning: {
		WorkflowRunEventPause:    WorkflowRunPaused,
		WorkflowRunEventComplete: WorkflowRunCompleted,
		WorkflowRunEventFail:     WorkflowRunFailed,
		WorkflowRunEventTimeout:  WorkflowRunTimedOut,
		WorkflowRunEventCancel:   WorkflowRunCanceled,
	},
	WorkflowRunPaused: {
		WorkflowRunEventResume:   WorkflowRunRunning,
		WorkflowRunEventComplete: WorkflowRunCompleted,
		WorkflowRunEventFail:     WorkflowRunFailed,
		WorkflowRunEventTimeout:  WorkflowRunTimedOut,
		WorkflowRunEventCancel:   WorkflowRunCanceled,
	},
}

var terminalWorkflowRunStatuses = map[WorkflowRunStatus]bool{
	WorkflowRunCompleted: true,
	WorkflowRunFailed:    true,
	WorkflowRunTimedOut:  true,
	WorkflowRunCanceled:  true,
}

// CanTransitionWorkflowRun checks whether a workflow run transition is valid.
func CanTransitionWorkflowRun(from WorkflowRunStatus, event WorkflowRunEvent) bool {
	events, ok := workflowRunTransitions[from]
	if !ok {
		return false
	}
	_, ok = events[event]
	return ok
}

// TransitionWorkflowRun returns the next status after applying the event.
func TransitionWorkflowRun(from WorkflowRunStatus, event WorkflowRunEvent) (WorkflowRunStatus, error) {
	events, ok := workflowRunTransitions[from]
	if !ok {
		return "", fmt.Errorf("invalid workflow run status %q", from)
	}
	next, ok := events[event]
	if !ok {
		return "", fmt.Errorf("invalid transition: %q + %q", from, event)
	}
	return next, nil
}

// IsTerminalWorkflowRunStatus returns true if the status is a terminal (final) state.
func IsTerminalWorkflowRunStatus(status WorkflowRunStatus) bool {
	return terminalWorkflowRunStatuses[status]
}

File: packages/python-sdk/strait/fsm/_workflow.py
"""Workflow run lifecycle finite state machine."""

from __future__ import annotations

from enum import StrEnum


class WorkflowRunStatus(StrEnum):
    PENDING = "pending"
    RUNNING = "running"
    PAUSED = "paused"
    COMPLETED = "completed"
    FAILED = "failed"
    TIMED_OUT = "timed_out"
    CANCELED = "canceled"


class WorkflowRunEvent(StrEnum):
    START = "START"
    PAUSE = "PAUSE"
    RESUME = "RESUME"
    COMPLETE = "COMPLETE"
    FAIL = "FAIL"
    TIMEOUT = "TIMEOUT"
    CANCEL = "CANCEL"


_workflow_run_transitions: dict[WorkflowRunStatus, dict[WorkflowRunEvent, WorkflowRunStatus]] = {
    WorkflowRunStatus.PENDING: {
        WorkflowRunEvent.START: WorkflowRunStatus.RUNNING,
        WorkflowRunEvent.CANCEL: WorkflowRunStatus.CANCELED,
    },
    WorkflowRunStatus.RUNNING: {
        WorkflowRunEvent.PAUSE: WorkflowRunStatus.PAUSED,
        WorkflowRunEvent.COMPLETE: WorkflowRunStatus.COMPLETED,
        WorkflowRunEvent.FAIL: WorkflowRunStatus.FAILED,
        WorkflowRunEvent.TIMEOUT: WorkflowRunStatus.TIMED_OUT,
        WorkflowRunEvent.CANCEL: WorkflowRunStatus.CANCELED,
    },
    WorkflowRunStatus.PAUSED: {
        WorkflowRunEvent.RESUME: WorkflowRunStatus.RUNNING,
        WorkflowRunEvent.COMPLETE: WorkflowRunStatus.COMPLETED,
        WorkflowRunEvent.FAIL: WorkflowRunStatus.FAILED,
        WorkflowRunEvent.TIMEOUT: WorkflowRunStatus.TIMED_OUT,
        WorkflowRunEvent.CANCEL: WorkflowRunStatus.CANCELED,
    },
}

_terminal_workflow_run_statuses: frozenset[WorkflowRunStatus] = frozenset({
    WorkflowRunStatus.COMPLETED,
    WorkflowRunStatus.FAILED,
    WorkflowRunStatus.TIMED_OUT,
    WorkflowRunStatus.CANCELED,
})


def can_transition_workflow_run(
    from_status: WorkflowRunStatus, event: WorkflowRunEvent,
) -> bool:
    events = _workflow_run_transitions.get(from_status)
    if events is None:
        return False
    return event in events


def transition_workflow_run(
    from_status: WorkflowRunStatus, event: WorkflowRunEvent,
) -> WorkflowRunStatus:
    events = _workflow_run_transitions.get(from_status)
    if events is None:
        raise ValueError(f"invalid workflow run status {from_status!r}")
    next_status = events.get(event)
    if next_status is None:
        raise ValueError(f"invalid transition: {from_status!r} + {event!r}")
    return next_status


def is_terminal_workflow_run_status(status: WorkflowRunStatus) -> bool:
    return status in _terminal_workflow_run_statuses
